geniza/corpus/models.py

```diff
@@ -764,7 +766,24 @@ def iiif_images(self, filter_side=False, with_placeholders=False):
                 for canvas_uri in ed.content_html.keys():
                     if canvas_uri not in iiif_images:
                         # use placeholder image for each canvas not in iiif_images
-                        iiif_images[canvas_uri] = Document.PLACEHOLDER_CANVAS
+                        iiif_images[canvas_uri] = deepcopy(Document.PLACEHOLDER_CANVAS)
+                        uri_match = re.search(
+                            r"textblock\/(\d+)\/canvas\/(\d)\/", canvas_uri
+                        )
+                        if uri_match:
+                            # if this was created using placeholders in the transcription editor,
+                            # try to ascertain and display the right fragment shelfmark and label
+                            try:
+                                iiif_images[canvas_uri][
+                                    "shelfmark"
+                                ] = TextBlock.objects.get(
+                                    pk=int(uri_match.group(1))
+                                ).fragment.shelfmark
```

Comment:
this looks like an extra db call per placeholder canvas

Reply:
you should be able to find the textblock from the current document object, which should avoid extra db calls. if you use `document.textblock_set.filter` I think it still hits the db hit, so maybe a list comprehension to get the right textblock by id.  (it does feel a bit weird but I get the logic)